Release 0.2.6 compile error with video

Hello @0xCCF4 ,

Thanks for the quick release of version 0.2.6.

Unfortunately compiling photo_sort with the video option doesn't work yet. `cargo install --features video photo_sort`

See the error below:
```
   Compiling filetime v0.2.26
   Compiling threadpool v1.8.1
   Compiling clap v4.5.47
   Compiling symlink v0.1.0
   Compiling photo_sort v0.2.6
error[E0308]: mismatched types
   --> /home/roy/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/photo_sort-0.2.6/src/lib.rs:232:48
    |
232 |             return self.analyze_video_metadata(path);
    |                         ---------------------- ^^^^ expected `&PathBuf`, found `&Path`
    |                         |
    |                         arguments to this method are incorrect
    |
    = note: expected reference `&PathBuf`
               found reference `&Path`
note: method defined here
   --> /home/roy/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/photo_sort-0.2.6/src/lib.rs:207:8
    |
207 |     fn analyze_video_metadata(&self, path: &PathBuf) -> Result<Option<NaiveDateTime>> {
    |        ^^^^^^^^^^^^^^^^^^^^^^        --------------

For more information about this error, try `rustc --explain E0308`.
error: could not compile `photo_sort` (lib) due to 1 previous error
error: failed to compile `photo_sort v0.2.6`, intermediate artifacts can be found at `/tmp/cargo-installrdG0Ik`.

```
Would you please fix this?

Small hint: https://github.com/0xCCF4/PhotoSort/issues/84

--- a/.github/workflows/check-compile.yaml
+++ b/.github/workflows/check-compile.yaml
@@ -1,0 +1,41 @@
+name: Install crates
+
+on:
+  schedule:
+    - cron: "* 13 * * 1"
+
+env:
+  RUSTFLAGS: -Dwarnings
+
+jobs:
+  crates_build:
+    strategy:
+      matrix:
+        os: [ ubuntu-latest ]
+        rustalias: [ stable, nightly ]
+        feature_flag: [ "--no-default-features", "", "--features video" ]
+        include:
+          - rustalias: stable
+            rust: stable
+          - rustalias: nightly
+            rust: nightly
+
+    name: 'Build and test ${{ matrix.feature_flag }}: ${{ matrix.os }}, ${{ matrix.rustalias }}'
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v5
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: ${{ matrix.rust }}
+          override: true
+
+      - name: Install ffmepg dependencies
+        run: sudo apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
+
+      - name: Install
+        uses: actions-rs/cargo@v1
+        with:
+          command: install
+          args: photo_sort ${{ matrix.feature_flag }}

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -19,37 +19,40 @@ jobs:
     if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name != github.event.pull_request.base.repo.full_name
     strategy:
       matrix:
-        os: [ubuntu-latest]
-        rustalias: [stable, nightly]
-        feature_flag: ["--no-default-features", ""]
+        os: [ ubuntu-latest ]
+        rustalias: [ stable, nightly ]
+        feature_flag: [ "--no-default-features", "", "--features video" ]
         include:
-        - rustalias: stable
-          rust: stable
-        - rustalias: nightly
-          rust: nightly
+          - rustalias: stable
+            rust: stable
+          - rustalias: nightly
+            rust: nightly
 
     name: 'Build and test ${{ matrix.feature_flag }}: ${{ matrix.os }}, ${{ matrix.rustalias }}'
     runs-on: ${{ matrix.os }}
     steps:
-    - uses: actions/checkout@v5
+      - uses: actions/checkout@v5
 
-    - name: Install Rust
-      uses: actions-rs/toolchain@v1
-      with:
-        toolchain: ${{ matrix.rust }}
-        override: true
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: ${{ matrix.rust }}
+          override: true
 
-    - name: Check
-      uses: actions-rs/cargo@v1
-      with:
-        command: check
-        args: --all ${{ matrix.feature_flag }} --bins --examples
+      - name: Install ffmepg dependencies
+        run: apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
 
-    - name: Tests
-      uses: actions-rs/cargo@v1
-      with:
-        command: test
-        args: --all ${{ matrix.feature_flag }}
+      - name: Check
+        uses: actions-rs/cargo@v1
+        with:
+          command: check
+          args: --all ${{ matrix.feature_flag }} --bins --examples
+
+      - name: Tests
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+          args: --all ${{ matrix.feature_flag }}
 
   cargo_fmt:
     if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name != github.event.pull_request.base.repo.full_name
@@ -71,24 +74,26 @@ jobs:
     if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name != github.event.pull_request.base.repo.full_name
     strategy:
       matrix:
-        feature_flag: ["--no-default-features", ""]
+        feature_flag: [ "--no-default-features", "", "--features video" ]
     runs-on: ubuntu-latest
     name: 'Style and docs ${{ matrix.feature_flag }}'
     steps:
-    - uses: actions/checkout@v5
-    - uses: actions-rs/toolchain@v1
-      with:
-        profile: minimal
-        toolchain: nightly
-        override: true
-        components: clippy
-    - name: clippy
-      uses: actions-rs/cargo@v1
-      with:
-        command: clippy
-        args: --all-targets ${{ matrix.feature_flag }} -- -D warnings
-    - name: docs
-      uses: actions-rs/cargo@v1
-      with:
-        command: doc
-        args: --no-deps ${{ matrix.feature_flag }}
+      - uses: actions/checkout@v5
+      - uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: nightly
+          override: true
+          components: clippy
+      - name: Install ffmepg dependencies
+        run: apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
+      - name: clippy
+        uses: actions-rs/cargo@v1
+        with:
+          command: clippy
+          args: --all-targets ${{ matrix.feature_flag }} -- -D warnings
+      - name: docs
+        uses: actions-rs/cargo@v1
+        with:
+          command: doc
+          args: --no-deps ${{ matrix.feature_flag }}

--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -40,7 +40,7 @@ jobs:
           override: true
 
       - name: Install ffmepg dependencies
-        run: apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
+        run: sudo apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
 
       - name: Check
         uses: actions-rs/cargo@v1
@@ -86,7 +86,7 @@ jobs:
           override: true
           components: clippy
       - name: Install ffmepg dependencies
-        run: apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
+        run: sudo apt install -y clang libavcodec-dev libavformat-dev libavutil-dev pkg-config ffmpeg libavutil-dev libavfilter-dev libavdevice-dev
       - name: clippy
         uses: actions-rs/cargo@v1
         with:

--- a/src/action.rs
+++ b/src/action.rs
@@ -3,7 +3,7 @@ use filetime::FileTime;
 use log::{debug, error, warn};
 use std::fmt::{Display, Formatter};
 use std::fs;
-use std::path::{Path, PathBuf};
+use std::path::Path;
 use std::str::FromStr;
 
 /// `ActualAction` is an enumeration that defines the different types of actions that can be performed on a file.
@@ -102,12 +102,15 @@ impl FromStr for ActualAction {
 ///
 /// * The target file already exists.
 /// * An error occurred during the file operation.
-pub fn file_action(
-    source: &PathBuf,
-    target: &PathBuf,
+pub fn file_action<P: AsRef<Path>, Q: AsRef<Path>>(
+    source: P,
+    target: Q,
     action: &ActionMode,
     mkdir: bool,
 ) -> Result<()> {
+    let source = source.as_ref();
+    let target = target.as_ref();
+
     error_file_exists(target)
         .map_err(|e| anyhow!("Target file already exists: {:?} - {:?}", target, e))?;
 

--- a/src/analysis/name_formatters/bracketed.rs
+++ b/src/analysis/name_formatters/bracketed.rs
@@ -30,10 +30,10 @@ impl NameFormatter for BracketedFormat {
             Some("?len"|"?length") => invocation_info.bracket_info.as_ref().map(|b|b.sequence_length.to_string()),
             Some("?group"|"?grp"|"?num") => invocation_info.bracket_info.as_ref().map(|b|b.group_index.to_string()),
             Some(x) => {
-                return Err(anyhow!("Unknown format argument {x:?} for the {{bracket}} format specifier. Possible values are \"seq/name/name_first/name_last/len/length/group/num\""))
+                return Err(anyhow!("Unknown format argument {x:?} for the {{bracket}} format specifier. Possible values are \"index/first/last/length/group\""))
             }
             None => {
-                return Err(anyhow!("No format argument for the {{bracket}} format string was specified. Possible values are \"seq/name/name_first/name_last/len/length/group/num\""));
+                return Err(anyhow!("No format argument for the {{bracket}} format string was specified. Possible values are \"index/first/last/length/group\""));
             }
         };
 

--- a/src/analysis/video2date.rs
+++ b/src/analysis/video2date.rs
@@ -33,7 +33,7 @@ fn init_ffmpeg() -> anyhow::Result<()> {
 /// # Errors
 /// This function will return an error if:
 /// * The video file could not be read.
-pub fn get_video_time<P: AsRef<Path> + ?Sized>(path: &P) -> anyhow::Result<Option<NaiveDateTime>> {
+pub fn get_video_time<P: AsRef<Path>>(path: P) -> anyhow::Result<Option<NaiveDateTime>> {
     init_ffmpeg()?;
 
     let instance = ffmpeg::format::input(&path)?;

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -204,7 +204,7 @@ impl Analyzer {
     }
 
     #[cfg(feature = "video")]
-    fn analyze_video_metadata(&self, path: &PathBuf) -> Result<Option<NaiveDateTime>> {
+    fn analyze_video_metadata<P: AsRef<Path>>(path: P) -> Result<Option<NaiveDateTime>> {
         let video_time = analysis::video2date::get_video_time(path)?;
         Ok(video_time)
     }
@@ -229,7 +229,7 @@ impl Analyzer {
         }
         #[cfg(feature = "video")]
         if video {
-            return self.analyze_video_metadata(path);
+            return Analyzer::analyze_video_metadata(path);
         }
 
         Err(anyhow::anyhow!("File extension is not valid"))
@@ -434,7 +434,12 @@ impl Analyzer {
     /// * An IO error occurs while analyzing the date
     /// * An IO error occurs while doing the file action
     #[allow(clippy::too_many_lines)]
-    pub fn run_file(&self, path: &PathBuf, bracket_info: &Option<BracketInfo>) -> Result<()> {
+    pub fn run_file<P: AsRef<Path>>(
+        &self,
+        path: P,
+        bracket_info: &Option<BracketInfo>,
+    ) -> Result<()> {
+        let path = path.as_ref();
         let valid_ext = self.is_valid_extension(path.extension());
         let is_unknown_file = match valid_ext {
             Ok(false) => {
@@ -495,7 +500,7 @@ impl Analyzer {
         }
         #[cfg(feature = "video")]
         if self.is_valid_video_extension(path.extension())? {
-            ftype = FileType::Video
+            ftype = FileType::Video;
         }
 
         let mut file_name_info = NameFormatterInvocationInfo {

